### PR TITLE
[plsql] added some additional info in userData

### PR DESCRIPTION
**Pull request by szyman23**
Before submitting a PR, please check that:
 - [ ] The PR is submitted against `master`. The PMD team will merge back to support branches as needed.
 - [ ] `./mvnw clean verify` passes. This will [build](https://github.com/pmd/pmd/blob/master/BUILDING.md) and test PMD, execute PMD and checkstyle rules. [Check this for more info](https://github.com/pmd/pmd/blob/master/CONTRIBUTING.md#code-style)

**PR Description:**
We added some additional info in userData so it can be used in rules:
- SqlStatement - COMMIT or ROLLBACK
- FormalParameterMode - IN, OUT, NO COPY
- FetchStatement - BULK COLLECT, BULK COLLECT LIMIT

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">No java rules are changed!</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-plsql/etc/grammar/PldocAST.jjt [hunk 1]
```diff
@@ -607,14 +607,22 @@ ASTFormalParameter FormalParameter()  :
...
+    (LOOKAHEAD (2) (<OUT>) FormalParameterMode())?
+    (LOOKAHEAD (3) (<NOCOPY>) FormalParameterMode())?
 	("..."    | Datatype())
 	( (":" "="|<_DEFAULT>) Expression() )?
    )
    { jjtThis.setImage(simpleNode.getImage()) ;  return jjtThis ; }
 }
 
-
+ASTFormalParameterMode FormalParameterMode():
```
Comment (adangel) [inline pmd-plsql/etc/grammar/PldocAST.jjt]: Wouldn't it be more useful to enhance [ASTFormalParameter](https://github.com/pmd/pmd/blob/master/pmd-plsql/src/main/java/net/sourceforge/pmd/lang/plsql/ast/ASTFormalParameter.java)?

You could define a enum there for the "Mode" and set it above in lines 610-612: `<IN> jjtThis.setMode(ASTFormalParameter.Mode.IN)` and so on....

We should also make sure to cover this in tests, I don't know, whether `x IN OUT NUMBER` is still correctly recognized...

The class ASTFormalParameters can then have several getters, e.g. `isIn()`, `isInOut()`, `isNoCopy()` ... which makes this easily consumable in a XPath rule.

I don't think, these tokens justify to create a complete new AST node...

File: pmd-plsql/etc/grammar/PldocAST.jjt [hunk 2]
```diff
@@ -1157,7 +1165,7 @@ ASTReadPastNextOccurrence ReadPastNextOccurrence(String target) :
 ASTSqlStatement SqlStatement(String initiator, String terminator)  :
 {}
 {
-	(<SELECT>|<UPDATE>|<INSERT>|<DELETE>|<COMMIT>|<ROLLBACK>|<SAVEPOINT>|<SET><TRANSACTION>|<LOCK><TABLE>|<MERGE>|<WITH>)
+    (<SELECT>|<UPDATE>|<INSERT>|<DELETE>|<COMMIT>{jjtThis.setUserData("COMMIT"); }|<ROLLBACK>{jjtThis.setUserData("ROLLBACK"); }|<SAVEPOINT>|<SET><TRANSACTION>|<LOCK><TABLE>|<MERGE>|<WITH>)
```
Comment (adangel) [inline pmd-plsql/etc/grammar/PldocAST.jjt]: We should create the class `ASTSqlStatement` and enhance this to set the statement there ... this makes it more accessible and useful to be used - setUserData is solely based on convention. In order to use it, you need to know that ASTSqlStatement stores a string there - but other nodes can store other data...

I'd suggest to create `setType(String)` and `getType()` method there.

Please also note, that this production doesn't actually parse SQL, it just **skips** the stuff completely...

File: pmd-plsql/etc/grammar/PldocAST.jjt [hunk 3]
```diff
@@ -2658,9 +2666,9 @@ ASTOpenStatement OpenStatement()  :
 ASTFetchStatement FetchStatement()  :
 {}
 {
-	 <FETCH> QualifiedName() [<BULK> <COLLECT>]
+	 <FETCH> QualifiedName() [<BULK> <COLLECT> {jjtThis.setUserData("BULK COLLECT"); }]
```
Comment (adangel) [inline pmd-plsql/etc/grammar/PldocAST.jjt]: We should add a boolean flag on the fetch statement: `jjtThis.setBulkCollect()` / `ASTFetchStatement.isBulkCollect()`.

**Comment by adangel** [comment]
Replaced by #2449

**Pull request closed without merging**